File: clase-07/python/main.py
def dividir_numeros():
    try:
        # El bloque try contiene el código que puede generar una excepción.
        print("Inicio del bloque try")          # Breakpoint 1

        # Si el usuario escribe algo que no se puede convertir a entero,
        # Python lanza una excepción ValueError.
        a = int(input("Ingrese el primer número: "))   # Breakpoint 2
        b = int(input("Ingrese el segundo número: "))  # Breakpoint 3

        # Si b vale 0, Python lanza una excepción ZeroDivisionError.
        resultado = a / b                       # Breakpoint 4
        print(f"Resultado: {resultado}")        # Breakpoint 5

    except ValueError:
        # Captura el error cuando el usuario no ingresa un número entero válido.
        print("Error: debe ingresar números enteros.")  # Breakpoint 6

    except ZeroDivisionError as e:
        # Captura el error cuando se intenta dividir por cero.
        print("Error: no se puede dividir por cero.")   # Breakpoint 7
        print(f"Texto del error:{e}")
    except Exception as e:
        # Excepción genérica:
        # captura cualquier otro error no contemplado en los except anteriores.
        # Conviene ponerla al final, porque si estuviera antes,
        # atraparía también a las excepciones específicas.
        print(f"Ocurrió un error no previsto:{e}")

    finally:
        # El bloque finally se ejecuta siempre:
        # - si todo salió bien
        # - si ocurrió un error
        # - si entró a un except
        # Se usa mucho para cerrar archivos, conexiones o liberar recursos.
        print("Fin del proceso.")               # Breakpoint 8


dividir_numeros()

File: clase-07/golang/main.go
package main

import (
	"errors"
	"fmt"
	"strconv"
)

func dividir(aStr string, bStr string) (float64, error) {
	// Esta función intenta convertir dos textos a número
	// y luego realizar una división.
	fmt.Println("Inicio de dividir()") // Breakpoint 1

	// strconv.Atoi convierte string a int.
	// Si el valor no es numérico, devuelve error.
	a, err := strconv.Atoi(aStr)
	if err != nil {
		// En Go no usamos try/catch:
		// devolvemos el error para que quien llama decida qué hacer.
		return 0, errors.New("el primer valor no es un entero válido") // Breakpoint 2
	}

	b, err := strconv.Atoi(bStr)
	if err != nil {
		return 0, errors.New("el segundo valor no es un entero válido") // Breakpoint 3
	}

	// Antes de dividir, validamos que el divisor no sea cero.
	if b == 0 {
		return 0, errors.New("no se puede dividir por cero") // Breakpoint 4
	}

	// Convertimos a float64 para evitar división entera.
	resultado := float64(a) / float64(b) // Breakpoint 5
	return resultado, nil
}

func main() {
	// Acá llamamos a la función y recibimos:
	// - el resultado
	// - el error (si ocurrió alguno)
	resultado, err := dividir("10", "0") // Breakpoint 6

	//resultado, err := dividir("10", "0") // Breakpoint 6

	if err != nil {
		// Si hubo error, lo mostramos y terminamos.
		fmt.Println("Error:", err) // Breakpoint 7
		return
	}

	// Si err == nil, significa que todo salió bien.
	fmt.Println("Resultado:", resultado) // Breakpoint 8
}
